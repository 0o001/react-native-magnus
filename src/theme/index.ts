export { ThemeContext } from './themeContext';
export { withTheme } from './withTheme';
export { ThemeProvider, ThemeProviderProps } from './themeProvider.component';

export {
  ThemeType,
  BorderPropsType,
  SpacingPropsType,
  RoundedPropsType,
} from './type';
