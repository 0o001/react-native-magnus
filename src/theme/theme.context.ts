import * as React from 'react';

import { ThemeType } from './type';

const defaultValue: ThemeType = {};

export interface ThemeContextType {
  theme: ThemeType;
  setTheme?: (theme?: ThemeType) => void;
}

export const THEME_CONTEXT_DEFAULT_VALUE = {
  theme: defaultValue,
};

export const ThemeContext: React.Context<ThemeContextType> = React.createContext(
  THEME_CONTEXT_DEFAULT_VALUE
);
