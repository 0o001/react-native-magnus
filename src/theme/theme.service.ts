import { Dimensions } from 'react-native';

import { ThemeType, BreakpointType } from './type';
import { defaultTheme } from '../style';

/**
 * merge user theme with default theme
 *
 * @param customTheme
 *
 * @return computedTheme computed theme variables object
 */
export function computeTheme(customTheme: ThemeType | undefined): ThemeType {
  const computedTheme = { ...defaultTheme };

  if (!customTheme) {
    return computedTheme;
  }

  if (customTheme.colors) {
    computedTheme.colors = {
      ...defaultTheme.colors,
      ...customTheme.colors,
    };
  }

  if (customTheme.fontSize) {
    computedTheme.fontSize = {
      ...defaultTheme.fontSize,
      ...customTheme.fontSize,
    };
  }

  if (customTheme.borderRadius) {
    computedTheme.borderRadius = {
      ...defaultTheme.borderRadius,
      ...customTheme.borderRadius,
    };
  }

  if (customTheme.name) {
    computedTheme.name = customTheme.name;
  }

  if (customTheme.breakpoints) {
    computedTheme.breakpoints = customTheme.breakpoints;
    computedTheme.breakpoint = calculateCurrentDeviceBreakpoint(
      customTheme.breakpoints
    );
  }

  return computedTheme;
}

/**
 * m="10"
 *
 * @param value
 * @param theme
 */
export const createSpacingStyles = (props: any, theme: any) => {
  let propKeys: any;

  propKeys = {
    p: 'padding',
    pt: 'paddingTop',
    pr: 'paddingRight',
    pb: 'paddingBottom',
    pl: 'paddingLeft',
    px: 'paddingHorizontal',
    py: 'paddingVertical',
    m: 'margin',
    mt: 'marginTop',
    mr: 'marginRight',
    mb: 'marginBottom',
    ml: 'marginLeft',
    mx: 'marginHorizontal',
    my: 'marginVertical',
    ms: 'marginStart',
  };

  const computedStyle: any = {};

  Object.keys(propKeys).map((propKey: string) => {
    const styleProperty = propKeys[propKey];

    if (propKey in props) {
      computedStyle[styleProperty] = getThemeProperty(
        theme,
        props[propKey],
        props[propKey]
      );
    }
  });

  return computedStyle;
};

/**
 * create directional style for properties like
 * top, left, right, bottom, horizontal, vertical
 *
 * @param key
 * @param value
 */
export const createDirectionalStyles = (
  key: string,
  value: any,
  type: string = 'string'
) => {
  if (
    typeof (value.x || value.y || value.b || value.t || value.l || value.r) ===
    'undefined'
  ) {
    return createStyle(`${key}`, value, type);
  }

  let computedStyle = {};

  if (typeof value.x !== 'undefined') {
    computedStyle = {
      ...computedStyle,
      ...createStyle(`${key}Horizontal`, value.x, type),
    };
  }

  if (typeof value.y !== 'undefined') {
    computedStyle = {
      ...computedStyle,
      ...createStyle(`${key}Vertical`, value.y, type),
    };
  }

  if (typeof value.t !== 'undefined') {
    computedStyle = {
      ...computedStyle,
      ...createStyle(`${key}Top`, value.t, type),
    };
  }

  if (typeof value.r !== 'undefined') {
    computedStyle = {
      ...computedStyle,
      ...createStyle(`${key}Right`, value.r, type),
    };
  }

  if (typeof value.b !== 'undefined') {
    computedStyle = {
      ...computedStyle,
      ...createStyle(`${key}Bottom`, value.b, type),
    };
  }

  if (typeof value.l !== 'undefined') {
    computedStyle = {
      ...computedStyle,
      ...createStyle(`${key}Left`, value.l, type),
    };
  }

  return computedStyle;
};

/**
 * extract the theme property from theme

 * @param theme
 * @param value
 */
export const getThemeProperty = (theme: any, value: any, fallback: any) => {
  if (typeof theme[value] !== 'undefined') {
    return theme[value];
  }

  return fallback;
};

/**
 * borderRadius="xl"
 *
 * @param value
 * @param theme
 */
export const createBorderRadiusStyles = (props: any, theme: any) => {
  let propKeys: any;

  propKeys = {
    rounded: 'borderRadius',
    roundedTopLeft: 'borderTopLeftRadius',
    roundedTopRight: 'borderTopRightRadius',
    roundedBottomLeft: 'borderBottomLeftRadius',
    roundedBottomRight: 'borderBottomRightRadius',
    roundedTop: ['borderTopLeftRadius', 'borderTopRightRadius'],
    roundedLeft: ['borderTopLeftRadius', 'borderBottomLeftRadius'],
    roundedRight: ['borderTopRightRadius', 'borderBottomRightRadius'],
    roundedBottom: ['borderBottomLeftRadius', 'borderBottomRightRadius'],
  };

  const computedStyle: any = {};

  Object.keys(propKeys).map((propKey: string) => {
    const styleProperty = propKeys[propKey];

    if (propKey in props) {
      if (styleProperty instanceof Array) {
        styleProperty.forEach((property) => {
          computedStyle[property] = getThemeProperty(
            theme,
            props[propKey],
            props[propKey]
          );
        });
      } else {
        computedStyle[styleProperty] = getThemeProperty(
          theme,
          props[propKey],
          props[propKey]
        );
      }
    }
  });

  return computedStyle;
};

/**
 * adds flex property in styles
 *
 * @param props
 */
export const createFlexStyles = (props: any) => {
  let propKeys: any;

  propKeys = {
    flex: 'flex',
  };

  const computedStyle: any = {};

  Object.keys(propKeys).map((propKey: string) => {
    const styleProperty = propKeys[propKey];

    if (propKey in props) {
      computedStyle[styleProperty] = props[propKey];
    }
  });

  return computedStyle;
};

/**
 * create borderWidth styles
 *
 * @param value
 */
export const createBorderWidthStyles = (props: any) => {
  let propKeys: any;

  propKeys = {
    borderWidth: 'borderWidth',
    borderTopWidth: 'borderTopWidth',
    borderEndWidth: 'borderEndWidth',
    borderLeftWidth: 'borderLeftWidth',
    borderRightWidth: 'borderRightWidth',
    borderBottomWidth: 'borderBottomWidth',
  };

  const computedStyle: any = {};

  Object.keys(propKeys).map((propKey: string) => {
    const styleProperty = propKeys[propKey];

    if (propKey in props) {
      computedStyle[styleProperty] = props[propKey];
    }
  });

  return computedStyle;
};

/**
 * create borderRadius styles
 *
 * @param value
 */
export const createBorderColorStyles = (props: any, theme: any) => {
  let propKeys: any;

  propKeys = {
    borderColor: 'borderColor',
    borderTopColor: 'borderTopColor',
    borderRightColor: 'borderTopColor',
    borderLeftColor: 'borderLeftColor',
    borderBottomColor: 'borderBottomColor',
  };

  const computedStyle: any = {};

  Object.keys(propKeys).map((propKey: string) => {
    const styleProperty = propKeys[propKey];

    if (propKey in props) {
      if (styleProperty instanceof Array) {
        styleProperty.forEach((property) => {
          computedStyle[property] = getThemeProperty(
            theme,
            props[propKey],
            props[propKey]
          );
        });
      } else {
        computedStyle[styleProperty] = getThemeProperty(
          theme,
          props[propKey],
          props[propKey]
        );
      }
    }
  });

  return computedStyle;
};

/**
 * adds shadows property in styles
 *
 * @param props
 */
export const createShadowStyles = (props: any, theme: any) => {
  let computedStyle: any = {};

  if (props.shadow) {
    computedStyle = {
      ...theme.shadow[props.shadow],
      shadowColor: getThemeProperty(theme.colors, props.shadowColor, 'white'),
    };
  }

  return computedStyle;
};

/**
 * create style object
 *
 * @param key
 * @param value
 */
export const createStyle = (key: string, value: any, type: string) => {
  return {
    [key]: type === 'number' ? Number(value) : value,
  };
};

/**
 *
 * @param props
 * @param computedStyle
 * @param styleProperty
 * @param atomicProperty
 * @param type
 */
export const addPropToComputedStyle = (
  props: any,
  computedStyle: any,
  styleProperty: string,
  atomicProperty: string,
  type: string
) => {
  if (props[atomicProperty]) {
    computedStyle = {
      ...computedStyle,
      ...createStyle(styleProperty, props[atomicProperty], type),
    };
  }

  return computedStyle;
};

/**
 * position="absolute"
 * top={10}
 *
 * @param value
 */
export const createPositionStyle = (props: any) => {
  let propKeys: any;

  propKeys = {
    position: 'position',
    top: 'top',
    left: 'left',
    right: 'right',
    bottom: 'bottom',
  };

  const computedStyle: any = {};

  Object.keys(propKeys).map((propKey: string) => {
    const styleProperty = propKeys[propKey];

    if (propKey in props) {
      computedStyle[styleProperty] = props[propKey];
    }
  });

  return computedStyle;
};

/**
 * position="absolute"
 * top={10}
 * w={[100%, 50%]}
 *  1005
 *  50
 * const DEFAULT_BREAKPOINTS = {
  sm: 320,
  md: 480,
  lg: 768,
  xl: 1024,
};
 *
 * @param value
 */
export const createResponsiveStyle = (theme: ThemeType, value: any) => {
  if (typeof value === 'string' || typeof value === 'number') {
    return value;
  }

  if (!theme.breakpoints) {
    throw new Error('No breakpoints defined');
  }

  if (!theme.breakpoint) {
    throw new Error('Not able to calculate current breakpoint');
  }

  const currentBreakpointValue = theme.breakpoints[theme.breakpoint];

  // if prop is array like bg=["red500", "blue500"]
  if (Array.isArray(value)) {
    if (value.length === 1) return value[0];

    let currentValue = value[0];

    if (currentBreakpointValue) {
      value.forEach((propValue, index) => {
        if (
          currentBreakpointValue >=
          theme.breakpoints[Object.keys(theme.breakpoints)[index]]
        ) {
          currentValue = propValue;
        } else return;
      });
    }

    return currentValue;
  }

  // if prop is object like bg={{ sm: 'red500', lg: 'blue500'}}
  if (typeof value === 'object') {
    let currentValue = value[Object.keys(value)[0]];

    Object.keys(value).forEach((breakpointKey, index) => {
      if (currentBreakpointValue >= theme.breakpoints[breakpointKey]) {
        currentValue = value[breakpointKey];
      } else return;
    });

    return currentValue;
  }

  return value;
};

/**
 * calculates current device breakpoint
 *
 * @param breakpoints
 */
export const calculateCurrentDeviceBreakpoint = (
  breakpoints: BreakpointType,
  currentWidth: number = Dimensions.get('window').width
) => {
  let currentBreakpoint = Object.keys(breakpoints)[0];

  Object.keys(breakpoints).forEach((breakpointKey: string) => {
    const breakpointValue = breakpoints[breakpointKey];

    if (currentWidth > breakpointValue) {
      currentBreakpoint = breakpointKey;
    } else return;
  });

  return currentBreakpoint;
};
