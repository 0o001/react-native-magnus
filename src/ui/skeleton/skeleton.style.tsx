import { StyleSheet } from 'react-native';

import {
  getThemeProperty,
  createSpacingStyles,
  createBorderRadiusStyles,
} from '../../theme/theme.service';

/**
 * computed style
 *
 * @param theme
 * @param props
 */
export const getStyle = (theme: any, props: any) => {
  const computedStyle: any = {};

  computedStyle.container = {
    ...createSpacingStyles(props, theme.spacing),
    backgroundColor: getThemeProperty(theme.colors, props.bg),
    ...createBorderRadiusStyles(props, theme.borderRadius),
  };

  if (props.flex) {
    computedStyle.container = {
      ...computedStyle.container,
      flex: props.flex,
    };
  }

  if (props.h) {
    computedStyle.container = {
      ...computedStyle.container,
      height: props.h,
    };
  }

  if (props.w) {
    computedStyle.container = {
      ...computedStyle.container,
      width: props.w,
    };
  }

  // merging style props to computed style
  if (props.style) {
    computedStyle.container = {
      ...computedStyle.container,
      ...props.style,
    };
  }

  return StyleSheet.create(computedStyle);
};
