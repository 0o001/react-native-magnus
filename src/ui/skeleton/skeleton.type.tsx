import { DivProps } from '../div/div.type';

export interface SkeletonProps extends DivProps {
  duration?: number;
}
