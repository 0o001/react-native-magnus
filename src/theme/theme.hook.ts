import React, { useState } from "react";

import { ThemeType } from "./type";
import { computeTheme } from "./theme.service";
import { ThemeContextType } from "./theme.context";

export const useTheme = (initialTheme?: ThemeType): ThemeContextType => {
  const [theme, setTheme] = useState(computeTheme(initialTheme));

  const setCurrentTheme = React.useCallback((newTheme?: ThemeType): void => {
    setTheme(computeTheme(newTheme));
  }, []);

  return {
    theme,
    setTheme: setCurrentTheme
  };
};
