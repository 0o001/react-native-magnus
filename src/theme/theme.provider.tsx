import * as React from "react";

import { ThemeType } from "./type";

import { ThemeContext } from "./theme.context";
import { useTheme } from "./theme.hook";

export interface ThemeProviderProps {
  theme?: ThemeType;
  children?: React.ReactNode;
}

export const ThemeProvider: React.FunctionComponent<ThemeProviderProps> = (
  props: ThemeProviderProps
) => {
  const { theme: themeProp, children } = props;
  const theme = useTheme(themeProp);

  return (
    <ThemeContext.Provider value={theme}>{children}</ThemeContext.Provider>
  );
};
