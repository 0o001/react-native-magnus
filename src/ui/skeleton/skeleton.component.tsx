import * as React from "react";
import { useContext } from "react";
import * as Animatable from "react-native-animatable";

import { getStyle } from "./skeleton.style";
import { ThemeContext } from "../../theme";
import { SkeletonProps } from "./skeleton.type";

const Box: React.FunctionComponent<SkeletonProps> = props => {
  const {
    m,
    mt,
    mr,
    mb,
    ml,
    p,
    pr,
    pt,
    pb,
    pl,
    bg,
    h,
    w,
    flex,
    rounded,
    roundedTop,
    roundedRight,
    roundedBottom,
    roundedLeft,
    ...rest
  } = props;
  const theme = useContext(ThemeContext);
  const computedStyle = getStyle(theme, props);

  Animatable.initializeRegistryWithDefinitions({
    fade: {
      0: {
        opacity: 1
      },
      0.5: {
        opacity: 0.3
      },
      1: {
        opacity: 1
      }
    }
  });

  return (
    <Animatable.View
      style={{ ...computedStyle.container }}
      animation="fade"
      duration={1000}
      easing="linear"
      iterationCount="infinite"
      {...rest}
    />
  );
};

const Circle: React.FunctionComponent<SkeletonProps> = props => {
  return <Box {...props} />;
};

Box.defaultProps = {
  bg: "gray400",
  h: 15,
  rounded: "md"
};

Circle.defaultProps = {
  bg: "gray400",
  h: 15,
  w: 15,
  rounded: "circle"
};

export { Box, Circle };
