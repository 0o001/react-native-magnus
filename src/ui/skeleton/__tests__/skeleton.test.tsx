import React from "react";
import renderer from "react-test-renderer";

import { Box } from "../skeleton.component";
import { ThemeProvider } from "../../../theme";

describe("Skeleton Component", () => {
  it("should render without issues", () => {
    const instance = renderer.create(
      <ThemeProvider>
        <Box />
      </ThemeProvider>
    );

    expect(instance.toJSON()).toMatchSnapshot();

    const typeIns = instance.root.findByType(Box);
    expect(typeIns.props.children).toBe(undefined);
  });
});
