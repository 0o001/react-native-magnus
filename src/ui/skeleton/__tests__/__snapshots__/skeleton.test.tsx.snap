// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Skeleton Component should render without issues 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#a6a6a6",
      "borderRadius": 6,
      "height": 15,
      "opacity": 1,
      "width": "100%",
    }
  }
/>
`;
